DatadogCore/Sources/Core/Upload/DataUploadWorker.swift

```diff
@@ -55,95 +63,124 @@ internal class DataUploadWorker: DataUploadWorkerType {
         self.contextProvider = contextProvider
         self.backgroundTaskCoordinator = backgroundTaskCoordinator
         self.delay = delay
+        self.maxBatchesPerUpload = maxBatchesPerUpload
         self.featureName = featureName
         self.telemetry = telemetry
 
-        let uploadWork = DispatchWorkItem { [weak self] in
+        createBatchReadWork()
+        scheduleBatchRead()
+    }
+
+    private func scheduleBatchRead() {
+        guard let batchReadWork = batchReadWork else {
+            return
+        }
+        queue.asyncAfter(deadline: .now() + delay.current, execute: batchReadWork)
+    }
+
+    private func createBatchReadWork() {
+        cancelUploads()
+        batchReadWork?.cancel()
+        batchReadWork = DispatchWorkItem { [weak self] in
             guard let self = self else {
                 return
             }
-
             let context = contextProvider.read()
             let blockersForUpload = self.uploadConditions.blockersForUpload(with: context)
             let isSystemReady = blockersForUpload.isEmpty
-            let nextBatch = isSystemReady ? self.fileReader.readNextBatch() : nil
-            if let batch = nextBatch {
+            let files = isSystemReady ? self.fileReader.readFiles(limit: maxBatchesPerUpload) : nil
+            if let files = files, !files.isEmpty {
+                DD.logger.debug("⏳ (\(self.featureName)) Uploading batches...")
                 self.backgroundTaskCoordinator?.beginBackgroundTask()
-                DD.logger.debug("⏳ (\(self.featureName)) Uploading batch...")
+                files.forEach { self.createAndScheduleBatchUploadWork(file: $0, context: context) }
+            } else {
+                let batchLabel = files?.isEmpty == false ? "YES" : (isSystemReady ? "NO" : "NOT CHECKED")
+                DD.logger.debug("💡 (\(self.featureName)) No upload. Batch to upload: \(batchLabel), System conditions: \(blockersForUpload.description)")
+                self.delay.increase()
+                self.backgroundTaskCoordinator?.endBackgroundTask()
```

Comment:
question/ It beings BG task each time when it finds a bach, but it ends it only when there are no more batches? How does it ensure that `begin()` is always matched with `end()`?

Reply:
It the same behaviour as the previous implementation. In the case mentioned it's balanced by the background task coordinator internal callback.